### PR TITLE
[Refactor] Add cross-platform Linking module

### PR DESCRIPTION
**Pull request by satya164**
A promise based API for handling Link for Android and iOS. Refer #4971 

The iOS part doesn't handle errors. Will need someone with iOS knowledge to do that.

cc @skevy @ide @brentvatne @mkonicek @vjeux @nicklockwood

### CODE REVIEW

**Comment by satya164** [comment]
We also need to decide what to do with `IntentAndroid` and `LinkingIOS`.

1. Show warnings stating that they are deprecated and use 'Linking', while they continue to work.
2. Show errors stating that they are deprecated and use 'Linking', making them noop.

**Comment by nicklockwood** [comment]
@satya164 

>We also need to decide what to do with IntentAndroid and LinkingIOS.

Option 1 is my preferred approach. Then in the next release we can maybe move to option 2, and then eventually remove them altogether.

**Comment by satya164** [comment]
@nicklockwood Awesome. I'll add the warnings.

**Comment by satya164** [comment]
@skevy @chirag04 Wanna help with the iOS implementation?

**Comment by satya164** [comment]
cc @dmmiller @nicklockwood

File: Libraries/Components/Intent/IntentAndroid.android.js [hunk 1]
```diff
@@ -119,24 +126,16 @@ class IntentAndroid {
    * it will give the link url, otherwise it will give `null`
    *
    * Refer http://developer.android.com/training/app-indexing/deep-linking.html#handling-intents
+   *
+   * @deprecated
    */
   static getInitialURL(callback: Function) {
+    console.warn('"IntentAndroid.getInitialURL" is deprecated. Use the promise based "Linking.getInitialURL" instead.');
```
Comment (mkonicek) [inline Libraries/Components/Intent/IntentAndroid.android.js]: nice!

**Comment by mkonicek** [comment]
Awesome! Thanks for working on this!

File: ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java [hunk 2]
```diff
@@ -85,9 +86,11 @@ public void openURL(String url) {
         intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
         getReactApplicationContext().startActivity(intent);
       }
+
+      promise.resolve(true);
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java]: I'm confused on what should I resolve with here.

File: Libraries/Linking/Linking.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+class Linking {
+  /**
+   * Add a handler to Linking changes by listening to the `url` event type
+   * and providing the handler
+   *
+   * @platform ios
+   */
+  static addEventListener(type: string, handler: Function) {
+    if (Platform.OS === 'android') {
+        console.error('Linking.addEventListener is not supported on Android');
```
Comment (satya164) [inline Libraries/Linking/Linking.js]: Maybe this can be just a warning, and not error?

**Comment by christopherdro** [comment]
Nice on @satya164! 
I'm wondering if we can work on getting https://github.com/facebook/react-native-applinks to work with this new API and have android support?

**Comment by mkonicek** [comment]
Had a quick look again, once you're happy with it feel free to shipit :)

**Comment by satya164** [comment]
@christopherdro I don't use that lib, but if someone can fix it to work with this change, it'll surely be great.

@mkonicek Only one doubt, what should I resolve with here - https://github.com/facebook/react-native/pull/5336/files#diff-fab623097f5131602a9aa8e4829fb6a9R90

null seems to be good candidate, but then I've no idea how to do the same with iOS - https://github.com/facebook/react-native/pull/5336/files#diff-07733d30787de00582b6e9693d4be698R83

**Comment by satya164** [comment]
Talked to @dmmiller . He's fine with `true`.

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Pull request closed without merging**